### ISSUE TITLE
Extend deepbots to support Evolutionary Algorithms

### ISSUE THREAD

**Issue by ManosMagnus** [aidudezzz/deepbots#86]
Initially deepbots was developed to support Reinforcement Learning algorithms however we expect that easily can be extended to support Evolutionary Algorithms. When it comes to evolutionary algorithm a population of agents are trained and mutated to solve a given task. At every episode the best agents are chosen to mutate in order to reach in a good enough solution. 

This project is quite open. We recommend to choose an easy task such as [Cartpole](https://github.com/aidudezzz/deepworlds/tree/dev/examples/cartpole) and adjust it on Evolutionary manner. We expect a grid of different agents that they try to solve the problem while the episodes are passed. We are open on using any evolutionary algorithm but we highly recommend to use a well established one. Finally, we expect to integrate the [Evolution-Guided Policy Gradient in Reinforcement Learning ](https://github.com/ShawK91/Evolutionary-Reinforcement-Learning) as proposed in [NIPS2018](https://papers.nips.cc/paper/2018/file/85fc37b18c57097425b52fc7afbb6969-Paper.pdf).

Any questions about what evolutionary algorithms can be uses, general questions or ideas are more than welcome!

**Comment by sanketsans** [comment]
Hi @ManosMagnus I am interested to work on this problem for this gsoc. 
**Contribution**

- Develop RL based environments which are based on openAI gym scenarios (​ https://gym.openai.com/envs/#classic_control​ ). I also want to develop other testbeds which are different from the gym based environments.
- The environments should be enhanced to support following categories :
○ continuous state, discrete action space.
○ Continuous state - action space.
○ Discrete space - action

- Implement and integrate DQN, CEM and REINFORCE algorithms in the codebase. Currently, deepbots support DDPG and PPO which are only based on continuous state-action pairs. Other algorithms supporting an enhanced environment catalog will help beginners and researchers to understand much better about the RL algorithms and how they work on different sets of state-action pairs.

**Comment by sanketsans** [comment]
I have previous experience working on creating environments and using them for training RL based agents. 

I have previous experience in building a python environment and then using it to train
a RL agent. ​ Dino AI​ (https://github.com/sanketsans/DinoAI) - where I created a simple clone of google chrome game.
I have 5 value based state objects and 3 actions ( 0 - nothing to do, 1 - jump, 2- crouch, it might vary for two agent scripts). The results are here.
https://www.youtube.com/watch?v=mrzyq8SkX0A

I have worked on the mentioned algorithms using openAI gym environment and maintain a github repo while applying the algorithms on different gym based environments. Check it here. ​ https://github.com/sanketsans/openAIenv

I am also currently in top leaderboards for some openai gym environments.

**Comment by veds12** [comment]
Hey @ManosMagnus I was looking into the algorithms that we could try to adjust the environments on and realised that there are quite a few of them. Do you have any preference as to which one of them I should start looking into first?

**Comment by ManosMagnus** [comment]
Hello @veds12, 
Indeed there are lots of genetic/evolutionary algorithm that can be used. As it is mentioned this project is quite open so feel free to share also your ideas in order to discuss them even further. 

Let me share some of my ideas:

First of all, I think that we can firstly try a simple genetic algorithm in order to setup the genetic training infrastructure. I find this [blog-post](https://machinelearningmastery.com/simple-genetic-algorithm-from-scratch-in-python/) quite helpful. On this way, we can build a training infrastructure that is based on genetic training (using the traditional mutation/crossover scheme). I totally recommend to use a tensor-based framework (such as [PyGAD](https://pypi.org/project/pygad/) and [pytorch](https://pytorch.org/)), which can be useful on future. Since the genetic algorithms follow the same philosophy, we can later easily  it with other approaches (such as Elitism, Adaptive GAs, CMA-ES). Some resources:

- [Medium article: Genetic Algorithm Implementation in Python](https://towardsdatascience.com/genetic-algorithm-implementation-in-python-5ab67bb124a6) 
- [Standard and Elitism genetic-algorithm repository](https://github.com/rmsolgi/geneticalgorithm)
- [Another GA repository](https://github.com/remiomosowon/pyeasyga)
- [PyGAD - this is might the most well established repository](https://github.com/ahmedfgad/GeneticAlgorithmPython)
- [Blog post using PyGAD library](https://neptune.ai/blog/adaptive-mutation-in-genetic-algorithm-with-python-examples)
-  [Adaptive Mutation in Genetic Algorithm with Python Examples using PyGAD](https://neptune.ai/blog/adaptive-mutation-in-genetic-algorithm-with-python-examples)
- [Covariance Matrix Adaptation Evolution Strategy](https://github.com/CMA-ES/pycma)
- Webots evolutionary algorithm sample: You can load the sample on webots: File > Open Sample World > samples > curriculum > advanced_genetic_algorithm.wbt

In my prospective, the first step could be the integration of PyGAD library on a simple example such as cartpole. In this way, we can create a wrapper infrastructure than can be on genetic training.

After that, the wrapper infrastructure can be extended to support Genetic Algorithm with Neural Network. In this case, we can also use PyGAD that supports GA with NN. Another algorithm could be also the [Neat](http://nn.cs.utexas.edu/downloads/papers/stanley.ec02.pdf) which is the classical algorithm for Evolving Neural Networks. Uber-Reasearch has some great resources for this task:

- [Welcoming the Era of Deep Neuroevolution](https://eng.uber.com/deep-neuroevolution/?ref=hackernoon.com)
- [uber-research/pytorch-neat](https://github.com/uber-research/PyTorch-NEAT)
- [uber-research/EvoGrad](https://github.com/uber-research/EvoGrad)
- [uber-research/deep-neuroevolution](https://github.com/uber-research/deep-neuroevolution)
- [ShawK91/Evolutionary-Reinforcement-Learning](https://github.com/ShawK91/Evolutionary-Reinforcement-Learning)
- [Design by Evolution: How to evolve your neural network with AutoML](https://www.kdnuggets.com/2017/07/design-evolution-evolve-neural-network-automl.html) 
- [Medium article: Genetic Algorithm with Neural Network](https://medium.com/@rajmandloi/genetic-algorithm-with-neural-network-cc9961221e42)

According to your preference you can develop a plan that can be feasible on given GSoC's timeline. Of course the about ideas are just that, ideas. If you come up with another idea or frameworks that can be used, we can discuss it as well.

**Comment by veds12** [comment]
Hey! Sorry for the late reply. I already went through some of this actually. I'll go through the rest of them in a couple of days and get back to you! We'll probably be able to have a better discussion then. Coming to frameworks, I was mostly planning to use PyTorch since I am very comfortable with that. I am not familiar PyGAD but seems worthwhile to look into it. I'll get back to you regarding this too, once I get an idea how it works 👍

**Comment by veds12** [comment]
Also, wanted to ask that how would the proposal phase work? Would this be the write to start formalising the ideas or would it be better to discuss and refine them first?

**Comment by ManosMagnus** [comment]
I feel that's both way are equally good. You should not leave it for the last day of deadline. I would recommend to develop a draft that formalize your idea and then discuss your idea in reference with your proposal. So feel free to send a draft proposal in the GFOSS's list in order to provide you feedback

**Comment by yiorgosynkl** [comment]
Hello @ManosMagnus !! The deepbots project ideas have sparked my interest (especially this one) and I have already been experimenting with openAI, webots and deepbots.
I am trying to strengthen my proposal and I was thinking that implementing a small project using deepbots would be a good idea. Do you encourage such an attempt? Do you have any suggestions?
Thanks in advance

**Comment by ManosMagnus** [comment]
Hello @yiorgosynkl, 

Of course we encourage such attempts. We will happy to help you. I will suggest to take a look on issue #27. We would like to extend deepbots in order to support a different kind of communication between `supervisor` and `robot`. That can happened via `customData`. I would recommend to replicate an easy example (such as [cartpole](https://github.com/aidudezzz/deepworlds/tree/dev/examples/cartpole/cartpole_discrete)) in order to test it. Feel free to ask for help in issue discussion

**Comment by Piyush-555** [comment]
Hi @ManosMagnus , if I'm not mistaken the goal of this project is to create examples that use Genetic Algorithms (popular algorithms on suitable environments) while keeping the implementation of the algorithm fairly general as done for DDPG and PPO in the deepworlds repository. And based on the situation, we may need to use emitter-receiver and/or robot-supervisor scheme. Is that correct?

**Comment by ManosMagnus** [comment]
Exactly! 

Let me elaborate this a bit. Let assume that we use the simplest possible example such as cartpole. First step might be to create a 3x3 grid in webots world of the 9 different instances of the cartpole problem (we called this as generation). Each of them (let call this robot) instance runs its own controller with the specific configuration. After the first generation the results (such as fitness value) of each robot should be send in a "central processing unit" (let call this supervisor). In turn, supervisor have to make the appropriate processing (such as mutations and combinations). After all, supervisor should reset the world as was in the first place and replace the new robots (new generation) in the scene by passing them the new configuration that is going to be used in generation. This should be repeated until a specific condition.

Currently, deepbots support only the one-to-one (supervisor-robot) communication. In genetic algorithms might be helpful to have a one-to-many (supervisor to many robots) communication scheme. That could be feasible either with emitter-receiver or with [customData](https://github.com/aidudezzz/deepbots/issues/27) field. Additionally, we need a way to formalize the scene setup. Finally, we need a abstraction or a generic way on supervisor controller to perform those functionalities (such as ` def create_new_generation(old_generation)` `def mutations(old_generatio)` etc). 

Of course this is just an idea, this can be also implemented in one by one manner. Both are welcome!

Please let me know if those make sense to you or if you have any concerns or improvements

**Comment by Piyush-555** [comment]
Sounds good!  
I think it is perfectly acceptable to modify deepbots API to support multi-agent communication, thus making the API more general.  
But do we need to create abstractions in deepbots API for GA-specific methods (like `create_new_generations`, `mutate`)? Won't this will be taken care of in the agent part of the implementation? Or did I misunderstood what you mean by 'abstraction'?

**Comment by yiorgosynkl** [comment]
Hello again!
When running the CarPole tutorial I came across this error message: 
![image](https://user-images.githubusercontent.com/56363368/113521128-269c1000-95a0-11eb-872b-42a3813904ce.png)

I tried to fix the error without any success. Do you have suggestions of how to solve this bug?

**Comment by veds12** [comment]
@yiorgosynkl you either need to set the python path to the environment where you install deepbots. Or launch webots from the command-line when the environment where you installed deepbots is active

**Comment by ManosMagnus** [comment]
@Piyush-555 

Yes this would be part of the agent. However, there are some "background" functionalities that the deepbots should implement. Simulation specific functionalities should be added on deepbots in order to provide an easy-to-use API to users. For example, after the and of every generation simulator should be reset at the initial state. I am not sure if mutate should be part of this API, but communication between supervisor and robots should be

**Comment by SidharajYadav** [comment]
I want to contribute 
please guide
